src/librbd/operation/SnapshotCreateRequest.cc

```diff
@@ -242,6 +241,46 @@ Context *SnapshotCreateRequest<I>::handle_create_object_map(int *result) {
   CephContext *cct = image_ctx.cct;
   ldout(cct, 5) << this << " " << __func__ << ": r=" << *result << dendl;
 
+  if (*result < 0) {
+    lderr(cct) << this << " " << __func__ << ": failed to snapshot object map: "
+               << cpp_strerror(*result) << dendl;
+
+    update_snap_context();
+    image_ctx.io_work_queue->unblock_writes();
+    return this->create_context_finisher(*result);
+  }
+
+  return send_create_image_state();
+}
+
+template <typename I>
+Context *SnapshotCreateRequest<I>::send_create_image_state() {
```

Comment:
Actually this is how I understood your suggestion (discussed via email) how to resolve this rase:

If it were in CreatePrimaryRequest, after create snapshot request is called (because we want to know snap_id), it would mean that snapshot create request would send image notification to refresh the image, so rbd mirror would refresh the image, saw the new primary snapshot, tried to read the image state object, but it could not exist at this stage yet. Thus we want to create the state object before we send the notification.

Reply:
Indeed ... forgot about that.